### DISCUSSION TITLE
Alternative to "pause_on_click_shorter_than"

### DISCUSSION THREAD

**Discussion by natural-harmonia-gropius** [tomasklaen/uosc#307] [Show and tell]
https://github.com/Natural-Harmonia-Gropius/mpv_config/blob/main/scripts/inputevent.lua

The timer can be inaccurate at heavy loads, so I think it's worth separating it to a separate script.

Example of input.conf
```ini
MBTN_LEFT           cycle pause                         #@click
MBTN_LEFT           cycle fullscreen                    #@double-click
MBTN_LEFT           show-text pressed                   #@press
MBTN_LEFT           show-text released                  #@release
MBTN_LEFT_DBL       ignore

SPACE               cycle pause                         #@click
SPACE               no-osd set speed 4; set pause no    #@press
SPACE               ignore                              #@release
SPACE               add volume -1                       #@repeat

PLAYPAUSE           cycle pause                         #@click
```

`*_DBL` need to be `ignore` from input.conf

**TODO:**

- [x] A good name, "on" is confusing. Current name from web development, onClick, ngOn, v-on...
  **inputevent.lua** now.
- [x] Make sure [executor](https://github.com/Natural-Harmonia-Gropius/mpv_config/blob/be8ab787e85de5c867017aa41a604f7b233606d2/scripts/on.lua#L87-L96) works properly.
- [x] Written with real OOP.
- [x] Support to bind in input.conf like `key command #@double_click`.
- [x] ~~Latency should be avoidable in some case.~~ nope.
  As low latency as possible now.
- [x] Auto restore when released as [pressaction](https://github.com/Natural-Harmonia-Gropius/mpv_config/blob/main/scripts/pressaction.lua)
- [x] ~~Use options.~~ Nothing to do.
- [x] ~~Drag the window will trigger the press.~~ Personally, I don't think it's a bug.
- [x] script-message bind/unbind.
  test.lua
  ```lua
  local utils = require("mp.utils")
  
  local key = "x"
  local on = {
      click = "cycle pause",
      double_click = "cycle fullscreen",
      press = "show-text pressed",
      release = "show-text released",
  }
  local json, err = utils.format_json(on)
  
  mp.add_forced_key_binding("z", "test", function()
      mp.commandv('script-message-to', 'inputevent', 'bind', key, json)
  end)
  ```
- [x] Other scripts may prevent events. Not sure is this can be solved.
  press and release works fine with `script-binding uosc/menu`.
- [x] Figure out repeat from mpv complex event.
  I don't know what the practical use of `repeat` is, ~~I tired `add volume -1` but mpv lower a lot of volume when I released.~~ debounced function 😪, be immdiatle now.
- [x] Figure out press from mpv complex event.
  ~~I didn't see the `press`.~~ `PLAYPAUSE` is press, replace it to click.
- [x] respect `--input-doubleclick-time`

@christoph-heinrich I don't know why `pause_on_click_shorter_than` is broken for me after the recent commit. It doesn't prevent click behavior on double click and drag causes dropped frames, maybe it's cycling pause repeatedly?

Can you take a look at this script and see what it could be improved?

**Comment by christoph-heinrich** [comment]
#309 should fix that functionality.

Also it is certainly possible to make a low latency solution, see https://github.com/christoph-heinrich/mpv-pointer-event.
It only has a delay after a single click to wait if it will become a double click. After the double click period, releasing the single click executes the `single_click` function immediately.

**Comment by natural-harmonia-gropius** [comment]
https://github.com/Natural-Harmonia-Gropius/InputEvent
It has its own repository, and I'm writing the documentation now.
